src/libprojectM/PCM.hpp

```diff
@@ -76,13 +76,16 @@ class Pcm
     static auto MakeTest() -> Test*;
 
 private:
-    template<int signalAmplitude, int signalOffset, class SampleType>
-    void AddPcm(
-        const SampleType* pcmData,
+    // CPP20: Could use a struct for the first 5 params to clarify on call site
+    // together with designated initializers
```

Comment:
As noted above, stride/offset should not be a compile-time argument, as it would not allow for passing 6-channel audio buffers to projectM, requiring apps to first reorganize their memory, which is not necessary as we copy the values into new buffers anyway.

Reply:
I don't think anyone should be passing 6-channel audio buffers into projectM. There's no good reason to do that. We should accept at most two channels.

Reply:
As discussed on Discord, almost all audio APIs use interleaved streams, so only allowing two-channel data to be passed in would mean to reorganize the data stream layout before passing it to projectM to [LR][LR]... format, just for projectM to copy each field into a separate array. By allowing any number of channels and specifying the (possibly optional) channel offsets in each frame, integrating apps don't need this additional step, and projectM can simply pick the proper values from the buffer and ignore the rest, without any additional cost for projectM _or_ the integrating app.

That said, most if not all audio APIs have a fixed channel order, which, for 8 channels, is: front left, front right, center, bass, rear left, rear right, side left, side right. Might be a bit different for unconventional formats like 2.1, but front L/R are always the first two and in this exact order. So we could provide two functions, one with default channel offsets (0 for left, 1 for right) and thus less parameters, and another one with user-defined offsets.

If applications need to downmix audio from 5.1 to stereo, for example, authors can still do that if required. But the API should make it seasy to pass the exact data the audio API provides as a pointer to projectM, and let the library pick the correct bytes from the buffer and perform any data type conversions like uint8->float.

Reply:
I think such an addition might be good to make its own task. I can see the value in it, but it feels a bit outside the scope of this one. The current API should be good for most uses as you said, and in a while we'll perhaps too be able to snoop how people make use of the library. At that point we'd also get some better data on where we could take API.

Reply:
Sure, no need to fix everything in this PR, just make it MVP.

Reply:
It's awesome that this is getting some love.  I've messed with these files, but wasn't brave enough to mess with the public APIs.  My one comment is that initial caps for method names seems very strange to me.  I've never worked a project that did that in C or Java, but maybe it's a common usage elsewhere...

Since I have messes with these files, here are some thoughts that may not have been shared:

I don't see any benefit to using doubles for PCM data.  32bit float is already very high precision for sound.  I've been annoyed by the conversion to double for rdft in the past.  I solved that in a different project (microprocessor) by just search/replace double->float.  That worked great.

One problem we haven't solved is that sometimes the audio data comes in too fast and we drop a lot of PCM data on the floor.  I think this is the cause of some complaints of "lag" when in fact we are actually way ahead of the "play head" rather than behind.  To fix that we'd need to use a much larger buffer and make use of the APIs like pa_stream_get_timing_info() (or equivalent in other libraries).  It's not a that much memory, but we need to buffer 0.5s of sound double the buffer size for no benefit seems kinda wasteful.

Reply:
I've seen issues with audio data buffering. If new data arrives in larger batches and the outside program just slams in the whole buffer, projectM throws away many samples due to the internal 2k ring buffer. This can result in multiple frames rendering the same wave data, beat detection not working properly and other weirdnesses.

I can't really think of a general solution for that problem we could implement in the library as the use case might be quite different. Audio players integrating projectM might provide samples way in advance to the library, which is also something Milkdrop has to take care of: see pluginshell.cpp, it asks Winamp for the actual delay to set the buffer size, then use a set number of samples per rendered frame to match the samplerate and player delay to use the exact samples currently playing. Then for standalone apps using recording devices you'll want a buffer that is more or less immediate, and fill it with the most current data before rendering the next frame to keep the visuals in sync as best as possible.

I think a configurable internal buffer size might be a good thing, so the external app can decide how many samples it should fill in advance, which can be calculated using sample rate times the expected delay.

Reply:
Thanks for clarifying my comments Kai.  Of course, as currently implemented the buffering problem is on the host side.  I had pondered adding a way to a) buffer more and b) add a way for the app to communicate position/latency information to PCM.  However, requiring the app to do isn't unreasonable.  BTW, the scenario that drove me nuts until I figured this out was Spotify on Linux which would cause the SDL demo app to send 0.5s of data at a time.